### ISSUE TITLE
endless loop with dproto

### ISSUE THREAD

**Issue by timotheecour** [dlang/dub#954]
problem occurs with "dproto": "~>2.1.0",
it goes away without this dependency.
it didn't use to cause problems

dub build --force --verbose

```
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Package ggplotd contains invalid dependency painlesstraits (ci=17)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Package dproto contains invalid dependency painlesstraits (ci=2)

```

with --vverbose:

```
  painlessjson (4, maybe optional): [1.3.7, 1.3.6, 1.3.5, 1.3.4, 1.3.3, 1.3.2, 1.3.1, invalid]
  color (24, required): [0.0.5]
  gtk-d (25, required): [3.3.1, 3.3.0]
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Dependency vibe-d:core -> libasync ~>0.7.8 mismatches with selected version 0.8.0 (ci=13)
Dependency cairod -> derelict-ft ~>1.0.2 mismatches with selected version 1.1.3 (ci=20)
Dependency derelict-ft -> derelict-util ~>2.0.6 mismatches with selected version 2.1.0 (ci=21)
Interation (ci=21) mir 0.17.0-alpha9, cpuid 0.2.6, dproto 2.1.0, painlesstraits invalid, painlessjson 1.3.7, dunit 1.0.12, msgpack-rpc 0.1.2, vibe-d 0.7.29, openssl 1.1.4+1.0.1g, botan 1.12.7, memutils 0.4.7, botan-math 1.0.3, libevent 2.0.1+2.0.16, libasync 0.8.0, libev 5.0.0+4.04, diet-ng 1.1.0, msgpack-d 0.9.6, ggplotd 1.1.0, dstats 1.0.3, cairod 0.0.1-alpha.3+1.10.2, derelict-ft 1.1.3, derelict-util 2.1.0 (C), windows-headers 1.0.1 (C), x11 1.0.15 (C), color 0.0.5 (C), gtk-d 3.3.1 (C)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Dependency vibe-d:core -> libasync ~>0.7.8 mismatches with selected version 0.8.0 (ci=13)
Dependency cairod -> derelict-ft ~>1.0.2 mismatches with selected version 1.1.3 (ci=20)
Interation (ci=20) mir 0.17.0-alpha9, cpuid 0.2.6, dproto 2.1.0, painlesstraits invalid, painlessjson 1.3.7, dunit 1.0.12, msgpack-rpc 0.1.2, vibe-d 0.7.29, openssl 1.1.4+1.0.1g, botan 1.12.7, memutils 0.4.7, botan-math 1.0.3, libevent 2.0.1+2.0.16, libasync 0.8.0, libev 5.0.0+4.04, diet-ng 1.1.0, msgpack-d 0.9.6, ggplotd 1.1.0, dstats 1.0.3, cairod 0.0.1-alpha.3+1.10.2, derelict-ft 1.1.3 (C), derelict-util 2.0.6 (C), windows-headers 1.0.1 (C), x11 1.0.15 (C), color 0.0.5 (C), gtk-d 3.3.1 (C)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Dependency vibe-d:core -> libasync ~>0.7.8 mismatches with selected version 0.8.0 (ci=13)
Dependency cairod -> derelict-ft ~>1.0.2 mismatches with selected version 1.1.2 (ci=20)
Interation (ci=20) mir 0.17.0-alpha9, cpuid 0.2.6, dproto 2.1.0, painlesstraits invalid, painlessjson 1.3.7, dunit 1.0.12, msgpack-rpc 0.1.2, vibe-d 0.7.29, openssl 1.1.4+1.0.1g, botan 1.12.7, memutils 0.4.7, botan-math 1.0.3, libevent 2.0.1+2.0.16, libasync 0.8.0, libev 5.0.0+4.04, diet-ng 1.1.0, msgpack-d 0.9.6, ggplotd 1.1.0, dstats 1.0.3, cairod 0.0.1-alpha.3+1.10.2, derelict-ft 1.1.2 (C), derelict-util 2.1.0 (C), windows-headers 1.0.1 (C), x11 1.0.15 (C), color 0.0.5 (C), gtk-d 3.3.1 (C)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Dependency vibe-d:core -> libasync ~>0.7.8 mismatches with selected version 0.8.0 (ci=13)
Dependency cairod -> derelict-ft ~>1.0.2 mismatches with selected version 1.1.1 (ci=20)
Interation (ci=20) mir 0.17.0-alpha9, cpuid 0.2.6, dproto 2.1.0, painlesstraits invalid, painlessjson 1.3.7, dunit 1.0.12, msgpack-rpc 0.1.2, vibe-d 0.7.29, openssl 1.1.4+1.0.1g, botan 1.12.7, memutils 0.4.7, botan-math 1.0.3, libevent 2.0.1+2.0.16, libasync 0.8.0, libev 5.0.0+4.04, diet-ng 1.1.0, msgpack-d 0.9.6, ggplotd 1.1.0, dstats 1.0.3, cairod 0.0.1-alpha.3+1.10.2, derelict-ft 1.1.1 (C), derelict-util 2.1.0 (C), windows-headers 1.0.1 (C), x11 1.0.15 (C), color 0.0.5 (C), gtk-d 3.3.1 (C)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Dependency vibe-d:core -> libasync ~>0.7.8 mismatches with selected version 0.8.0 (ci=13)
Dependency cairod -> derelict-ft ~>1.0.2 mismatches with selected version 1.1.0 (ci=20)
Interation (ci=20) mir 0.17.0-alpha9, cpuid 0.2.6, dproto 2.1.0, painlesstraits invalid, painlessjson 1.3.7, dunit 1.0.12, msgpack-rpc 0.1.2, vibe-d 0.7.29, openssl 1.1.4+1.0.1g, botan 1.12.7, memutils 0.4.7, botan-math 1.0.3, libevent 2.0.1+2.0.16, libasync 0.8.0, libev 5.0.0+4.04, diet-ng 1.1.0, msgpack-d 0.9.6, ggplotd 1.1.0, dstats 1.0.3, cairod 0.0.1-alpha.3+1.10.2, derelict-ft 1.1.0 (C), derelict-util 2.1.0 (C), windows-headers 1.0.1 (C), x11 1.0.15 (C), color 0.0.5 (C), gtk-d 3.3.1 (C)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)
Dependency vibe-d:core -> libasync ~>0.7.8 mismatches with selected version 0.8.0 (ci=13)
Package ggplotd contains invalid dependency painlesstraits (ci=17)
Interation (ci=17) mir 0.17.0-alpha9, cpuid 0.2.6, dproto 2.1.0, painlesstraits invalid, painlessjson 1.3.7, dunit 1.0.12, msgpack-rpc 0.1.2, vibe-d 0.7.29, openssl 1.1.4+1.0.1g, botan 1.12.7, memutils 0.4.7, botan-math 1.0.3, libevent 2.0.1+2.0.16, libasync 0.8.0, libev 5.0.0+4.04, diet-ng 1.1.0, msgpack-d 0.9.6, ggplotd 1.1.0 (C), dstats 1.0.3 (C), cairod 0.0.1-alpha.3+1.10.2 (C), derelict-ft 1.0.2 (C), derelict-util 2.1.0 (C), windows-headers 1.0.1 (C), x11 1.0.15 (C), color 0.0.5 (C), gtk-d 3.3.1 (C)
Package dproto contains invalid dependency painlesstraits (ci=2)
Package painlessjson contains invalid dependency painlesstraits (ci=4)

```

@s-ludwig please help, this is a blocker, can't even go back to a sane state anymore

**Comment by s-ludwig** [comment]
I suspect that this isn't really an endless loop, but that it hits an exponential case, where it has to try every combination of dependency versions to discover that there is no valid combination. It would be interesting to see if there is an easy way to detect this locally and error out early.

The actual problem appears to be in dproto, where the dependency to painlesstraits is `~>0.0.3`, while it is `~>0.2.0` in painlessjson, which has no matched version in common. A workaround would be to add a `"painlessjson": "~>1.3.5"` dependency to the root package, or at least that is the last version without the painlesstraits dependency.

**Comment by timotheecour** [comment]
are you sure there's no endless loop (which could indicate a bug in dub code)? it was running for many minutes until I killed it
Thanks for explaining the underlying mechanism (`has to try every combination of dependency versions to discover that there is no valid combination`), I didn't realize it was doing that; is that explained in docs?
Yes, there should be a way to detect statically by forming a graph of dependencies, but I suspect it does indicate a potential bug in dub (endless loop when dependencies are incompatible)

This is a nasty bug though for users not familiar with internals of dub; it's very likely to come back in the future, so the workaround you suggest could be ok temporarily, but it should provide a good test case to debug this issue IMO

Thanks for the reply!

**Comment by s-ludwig** [comment]
This is unfortunately a fundamental issue that cannot be solved in general. Dependency resolution is an exponential problem at its core. What we can do are two basic things:

1. Try to provide and hit fast paths as often as possible. This still has the risk of sometimes hitting the slow path.
2. Cancel the resolution process after a defined amount of maximum iterations. This can lead to valid dependency configurations being flagged as invalid.
3. Use an approximation instead of the exact solution. A polynomial algorithm could be used to yield the correct answer in most cases, with the risk of not being able to satisfy all dependency configurations.

The current and planned strategy is 1, with 2 being an idea worth considering. See also #652.

**Comment by timotheecour** [comment]
[did you get my email? would like to discuss about this]

adding to root package
`"painlessjson": "~>1.3.5",` 
didn't help

**Comment by s-ludwig** [comment]
Sorry, I used `~>` by habit. Should be just `"painlessjson" : "1.3.5"` (followed by a "dub upgrade" to actually change the version selection).

I got the e-mail, but I didn't get a free minute because I'm still trying to catch up before I need to resume some contract work. Maybe we can use the #vibe.d channel on freenode, though, it's low traffic, so if I don't catch something immediately it won't get lost.
